Support `@argumentDefinitions` in `@inline` fragments

tbezman opened the pull request:
This PR is the followup to #3933. Now we support both query & fragment variables inside of `@inline` directives.

Fixes #3106 

### What changed?

- Added `args` to the `InlineDirectiveMetadata` struct
- Piped those args through to the AST in `build_ast`
- Updated `RelayReader` to temporarily swap out the variables while traversing an inline fragment

daniloab commented:
Wow, awesome!

captbaritone commented on packages/relay-runtime/store/__tests__/readInlineData-test.js:
```diff
@@ -63,16 +64,120 @@ test('unwrap inline fragment data', () => {
 });
 
 describe('integration test with variables', () => {
+  test.only('works with fragment variables & query variables', () => {
```
```suggestion
  test('works with fragment variables & query variables', () => {
```

captbaritone commented on packages/relay-runtime/store/RelayReader.js:
```diff
@@ -1111,11 +1111,35 @@ class RelayReader {
     const inlineData = {};
     const parentFragmentName = this._fragmentName;
     this._fragmentName = fragmentSpreadOrFragment.name;
+
+    let temporaryTraversalVariables = this._variables;
+
+    // If the inline fragment spread has arguments, we need to temporarily
+    // switch this._variables to include the fragment spread's arguments
+    // for the duration of its traversal.
+    if (fragmentSpreadOrFragment.args) {
+      const spreadArguments = getArgumentValues(
+        fragmentSpreadOrFragment.args,
+        this._variables,
+      );
+
+      // Inherit the operation's arguments and give local fragment arguments higher priority
+      temporaryTraversalVariables = {...this._variables, ...spreadArguments};
+    }
+
+    // Swap out the variables
+    let oldVariables = this._variables;
+    this._variables = temporaryTraversalVariables;
+
     this._traverseSelections(
       fragmentSpreadOrFragment.selections,
       record,
       inlineData,
     );
+
+    // Put the old variables back
+    this._variables = oldVariables;
+
```
```suggestion

    let oldVariables = this._variables;
    
    // If the inline fragment spread has arguments, we need to temporarily
    // switch this._variables to include the fragment spread's arguments
    // for the duration of its traversal.
    if (fragmentSpreadOrFragment.args) {
      const spreadArguments = getArgumentValues(
        fragmentSpreadOrFragment.args,
        this._variables,
      );

      // Inherit the operation's arguments and give local fragment arguments higher priority
      this._variables = {...this._variables, ...spreadArguments};
    }
    
    this._traverseSelections(
      fragmentSpreadOrFragment.selections,
      record,
      inlineData,
    );

    // Put the old variables back
    this._variables = oldVariables;

```

captbaritone commented on packages/relay-runtime/util/ReaderNode.js:
```diff
@@ -33,6 +33,7 @@ export type ReaderInlineDataFragmentSpread = {|
   +kind: 'InlineDataFragmentSpread',
   +name: string,
   +selections: $ReadOnlyArray<ReaderSelection>,
+  +args?: ?$ReadOnlyArray<ReaderArgument>,
```
I think ideally this would be:

```suggestion
  +args: ?$ReadOnlyArray<ReaderArgument>,
```

But having it optional is a nice way to make it safe to land runtime and compiler changes separately internally. Maybe add a comment that we should come back and make the property non-optional (the value can continue to be optional).

tbezman replied:
@captbaritone so we'll basically follow this pr up right after with the non-optional?

captbaritone replied:
Yeah, that would be perfect.

captbaritone replied:
Actually, I'll do this internally.

captbaritone reviewed (changes requested):
This looks great overall! Just one issue with some debug stuff still left in the tests, and one style suggestion for the runtime change.

captbaritone commented:
Also, if you can rebase on `main` you should see some ESLint/Prettier/rustfmt errors coming from CI. I've made a few PRs which allows those to run on external PRs similarly to how they run on internal PRs.

tbezman commented on packages/relay-runtime/store/RelayReader.js:
```diff
@@ -1111,11 +1111,31 @@ class RelayReader {
+      );
+
+      // Inherit the operation's arguments and give local fragment arguments higher priority
+      this._variables = {...this._variables, ...spreadArguments};
```
@captbaritone is the idea here to expose `argumentDefinitions` on the `InlineDataFragmentSpread` so we can re-use that function?

```diff
return {
  "fragment": {
    "argumentDefinitions": (v0/*: any*/),
    "kind": "Fragment",
    "metadata": null,
    "name": "readInlineDataTestFragmentAndQueryVariablesQuery",
    "selections": [
      {
        "alias": null,
        "args": null,
        "concreteType": "User",
        "kind": "LinkedField",
        "name": "me",
        "plural": false,
        "selections": [
          {
            "kind": "InlineDataFragmentSpread",
            "name": "readInlineDataTestFragmentAndQueryVariables",
+           "argumentDefinitions": GET THESE HERE SOME HOW
            "selections": [
              {
                "alias": "fragmentVariable",
                "args": [
                  {
                    "kind": "Variable",
                    "name": "scale",
                    "variableName": "theScale"
                  }
                ],
                "concreteType": "Image",
                "kind": "LinkedField",
                "name": "profile_picture",
                "plural": false,
                "selections": (v1/*: any*/),
                "storageKey": null
              },
              (v2/*: any*/)
            ],
            "args": [
              {
                "kind": "Variable",
                "name": "theScale",
                "variableName": "scale2"
              }
            ]
          }
        ],
        "storageKey": null
      }
    ],
```

captbaritone replied:
I believe so. I think the `argumentDefinitions` include information about default values, which I believe is required to correctly compute the actual value of the arguments.

captbaritone commented:
Let me know if the rustfmt CI job is causing you grief. I just added it. I don't yet know a good way to enforce that folks developing locally are using the expected version of rustfmt.

tbezman commented:
@captbaritone no worries, yeah I can't seem to align my vscode with the version y'all are using. i think it's a noob thing but running `rustfmt {path/to/file}` always does the trick for me.

captbaritone commented:
Is this ready for another round? No hurry, just let me know when it is.

tbezman commented:
@captbaritone yup!

captbaritone commented on packages/relay-runtime/store/RelayReader.js:
```diff
@@ -1111,11 +1113,37 @@ class RelayReader {
+        this._variables,
+      );
+
+      RelayConcreteVariables.getFragmentVariables(
+        fragmentSpreadOrFragment,
+        this._variables,
+        argumentVariables,
+      );
+
+      // Inherit the operation's arguments and give local fragment arguments higher priority
+      this._variables = {...this._variables, ...argumentVariables};
```
```suggestion
       // Inherit the operation's arguments and give local fragment arguments higher priority
      this._variables = RelayConcreteVariables.getFragmentVariables(
        fragmentSpreadOrFragment,
        this._variables,
        argumentVariables,
      );
```

captbaritone commented on packages/relay-runtime/store/__tests__/readInlineData-test.js:
```diff
@@ -63,16 +64,120 @@ test('unwrap inline fragment data', () => {
+    const Fragment = graphql`
+      fragment readInlineDataTestFragmentAndQueryVariables on User
+      @inline
+      @argumentDefinitions(theScale: {type: "Float"}) {
```
Can we add a test that includes a fragment argument with a default value?

captbaritone commented on compiler/crates/relay-compiler/tests/compile_relay_artifacts/fixtures/inline-data-fragment-local-args.expected:
```diff
@@ -19,53 +18,191 @@ fragment inlineDataFragmentLocalArgsProfile on User
   }
 }
-==================================== ERROR ====================================
-✖︎ Variables from @argumentDefinitions are not yet supported inside @inline fragments.
```
So exciting! Love seeing "not supported" tests turning green.

captbaritone commented on compiler/crates/graphql-ir/src/ir.rs:
```diff
@@ -326,7 +326,7 @@ impl Condition {
 
 /// @ Name Arguments?
-#[derive(Clone, Debug, Eq, PartialEq)]
+#[derive(Clone, Debug, Eq, PartialEq, Hash)]
```
Just out of curiosity. Why do we need to derive `Hash` for these structs?

tbezman replied:
Started at `InlineDirectiveMetadata` needing `Hash`. Going down the chain from `InlineDirectiveMetadata` => `VariableDefinition` => `Directive`

captbaritone reviewed (changes requested):
Just one change in Reader which I suspect was just a simple oversight, and then a test to validate that default values work.

tbezman commented:
@captbaritone we found a bug at Coinbase. Query => Regular Fragment => Inline Fragment. I was using `this._variables` instead of `this._owner.variables`. This is now inline with how `getSingularSelector` works. I added a unit test to confirm the bug and fixed the issue. Feeling really good about this now.

captbaritone commented on packages/relay-runtime/store/RelayReader.js:
```diff
@@ -1111,11 +1113,40 @@ class RelayReader {
+    // If the inline fragment spread has arguments, we need to temporarily
+    // switch this._variables to include the fragment spread's arguments
+    // for the duration of its traversal.
+    if (fragmentSpreadOrFragment.args) {
+      const argumentVariables = getArgumentValues(
+        fragmentSpreadOrFragment.args,
+        this._variables,
+      );
+
+      this._variables = RelayConcreteVariables.getFragmentVariables(
+        fragmentSpreadOrFragment,
+        this._owner.variables,
+        argumentVariables,
+      );
+    } else {
+      this._variables = RelayConcreteVariables.getFragmentVariables(
+        fragmentSpreadOrFragment,
+        this._owner.variables,
+        {},
+      );
+    }
```
Style nit: Would this work?
```suggestion
    const argumentVariables = fragmentSpreadOrFragment.args ? getArgumentValues(
      fragmentSpreadOrFragment.args,
      this._variables,
    ) : {};

    this._variables = RelayConcreteVariables.getFragmentVariables(
      fragmentSpreadOrFragment,
      this._owner.variables,
      argumentVariables,
    );
```

captbaritone reviewed (changes requested):
One style nit (sorry!)

captbaritone commented:
(Flow CI error is not caused by this diff)

tbezman commented:
alright @captbaritone . looking good?
